### PR TITLE
ENH: change buffer format string for datetime64 to '=Q'

### PR DESCRIPTION
**Pull request by mattip**
Builds on PR #11694 for arrays. Add tests and fix implementation of
 `np.array(memoryview(np.array(3, 'M8[D]')))` 

Note this changes the format string for buffers created from `datetime64` and `timedelta64`,
from `B` with `itemsize=8` to `=Q` with `itemsize=1`. The array returned from the roundtrip
no longer has the `'M8[D]'` dtype, but at least it does not raise an exception.

### CODE REVIEW

**Comment by mattip** [comment]
Incorporated the review from #11694 for string formatting and using `=Q` instead of `Q`

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -1390,10 +1389,12 @@ _array_from_buffer_3118(PyObject *memoryview)
...
             /* This object has no excuse for a broken PEP3118 buffer */
-            PyErr_SetString(
+            PyErr_Format(
                     PyExc_RuntimeError,
-                    "Item size computed from the PEP 3118 buffer format "
-                    "string does not match the actual item size.");
+                    "Item size %zd for PEP 3118 buffer format "
+                    "string %s does not match the dtype %c item size %zd.",
+                    view->itemsize, view->format, descr->type,
+                    descr->elsize);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: This change would be nice to have in 1.15, especially since we hit this code path more

Comment (charris) [reply numpy/core/src/multiarray/ctors.c]: Wrong format for `elsize`, which is `int`.

**Comment by charris** [comment]
Not sure about the `B` change mentioned in test, I don't see it.

**Comment by charris** [comment]
Note that I pushed some fixes.

**Comment by charris** [comment]
Needs a rebase.

**Comment by eric-wieser** [comment]
How does this handle `dtype('<m8[s]')` and `dtype('>m8[s]')`? I feel like the `=` is already inserted somewhere else.

**Comment by eric-wieser** [comment]
An alternative proposal - make the typestring `T{Q:__value__:0B:__dtype_M8s}` or similar, using the name of an empty field as a side channel to pass our dtype

**Comment by mattip** [comment]
> Make the typestring T{Q:__value__:0B:__dtype_M8s:} or similar, using the name of an empty field as a side channel to pass our dtype

Interesting idea. Is there a precedent for using a side channel like that somewhere?

**Comment by mattip** [comment]
Closing this in favor of a future PR using a more descriptive type annotation like `T{Q:__value__:0B:__dtype_M8s:}` for `memoryview(datetime64)`

**Comment by aldanor** [comment]
> Closing this in favor of a future PR using a more descriptive type annotation like `T{Q:__value__:0B:__dtype_M8s:}` for `memoryview(datetime64)`

Wondering whether it was decided to discard this idea?

**Comment by seberg** [comment]
Ideas rarely get fully discarded :).  But I do not think that anyone is pursuing it or likely to do so in the foreseeable future and it may need a careful proposal.  Also in many cases `__array__` probably just solves the problem reasonably well (although of course it is not as generic as the buffer interface).

**Pull request closed without merging**